### PR TITLE
feat: support metadata filtering for the search endpoint

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

Enables metadata filtering for `/api/v1` search endpoint.

Also, this PR includes some code refactoring already done for the Similarity search feature (See https://github.com/argilla-io/argilla/pull/3815) to simplify the future merge commit from that branch.

Refs #3748 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor (change restructuring the codebase without changing functionality)
- [x] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

Some manual tests have been done in a local environment

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

**Comment by frascuchon** [comment]
@gabrielmbmb I will refactor the current param schemas to use this https://github.com/argilla-io/argilla/pull/3799/commits/6a155f593ca015292511f4ce0ff0381c71ecb0fd in a separate PR

File: src/argilla/server/apis/v1/handlers/datasets.py [hunk 1]
```diff
@@ -330,6 +332,36 @@ async def create_dataset_metadata_property(
...
+
+    # TODO: We should split API v1 into different FastAPI apps so we can customize error management.
+    # After mapping ValueError to 422 errors for API v1 then we can remove this try except.
+    try:
+        metadata_property = await datasets.create_metadata_property(db, dataset, metadata_prop_create)
+        return metadata_property
+    except ValueError as err:
+        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(err))
+
+
```
Comment (gabrielmbmb) [inline src/argilla/server/apis/v1/handlers/datasets.py]: I think this got duplicated when merging

File: src/argilla/server/search_engine/base.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        to: Optional[float]
+
+    def __post_init__(self):
+        if self.low is None and self.high is None:
+            raise ValueError("One of `low` or `high` value must be provided")
+
+    @classmethod
+    def from_string(cls, metadata_property: MetadataProperty, string: str) -> "FloatMetadataFilter":
+        model = cls.RangeModel.parse_raw(string)
+        return cls(metadata_property, model.from_, model.to)
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/base.py]: maybe we could simplify this if `{Terms,Integer,Float}MetadataFilter` were `BaseModel` directly

Comment (gabrielmbmb) [reply src/argilla/server/search_engine/base.py]: but we can tackle this in another PR

**Review by gabrielmbmb — CHANGES_REQUESTED** [review]

File: src/argilla/server/search_engine/base.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    def __post_init__(self):
+        if self.low is None and self.high is None:
+            raise ValueError("One of `low` or `high` value must be provided")
+
+    @classmethod
+    def from_string(cls, metadata_property: MetadataProperty, string: str) -> "IntegerMetadataFilter":
+        model = cls.RangeModel.parse_raw(string)
+        return cls(metadata_property, model.from_, model.to)
+
+
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/base.py:96]:
```suggestion
# TODO: transform to `pydantic.BaseModel`
```

File: src/argilla/server/search_engine/base.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+@dataclasses.dataclass
+class TermsMetadataFilter(MetadataFilter):
+    values: List[str]
+
+    @classmethod
+    def from_string(cls, metadata_property: MetadataProperty, string: str) -> "MetadataFilter":
+        return cls(metadata_property=metadata_property, values=string.split(","))
+
+
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/base.py:76]:
```suggestion
TODO: transform to `pydantic.BaseModel`
```

**Review by gabrielmbmb — CHANGES_REQUESTED** [review]

File: src/argilla/server/apis/v1/handlers/datasets.py (hunk 1 above)
Comment (frascuchon) [reply src/argilla/server/apis/v1/handlers/datasets.py]: YEs. Fixed!

**Pull request merged**